src/vmemcache_repl.c

```diff
@@ -338,44 +345,88 @@ repl_p_lru_evict(struct repl_p_head *head, struct repl_p_entry **ptr_entry)
 		goto exit_unlock;
 	}
 
-	if (ptr_entry == NULL) {
-		is_LRU = 1;
-		ptr_entry = TAILQ_FIRST(&head->first)->ptr_entry;
+	if (is_LRU) {
+		entry = TAILQ_FIRST(&head->first);
+		ptr_entry = entry->ptr_entry;
+	} else {
+		entry = *ptr_entry;
 	}
 
 	/*
 	 * Try to lock the entry by setting 'ptr_entry' to NULL,
 	 * so that it cannot be used nor evicted in other threads.
 	 */
-	entry = *ptr_entry;
-	if (entry == NULL ||
-	    !util_bool_compare_and_swap64(ptr_entry, entry, NULL)) {
+	if (entry != NULL && util_bool_compare_and_swap64(ptr_entry,
+								entry, NULL))
+		goto evict_found_entry;
 
-		/*
-		 * The first try failed. The entry can be locked
-		 * and enqueued in the ring buffer, so let's flush
-		 * the ring buffer and try again.
-		 */
-		dequeue_all(head);
+	/*
+	 * The first try failed. The entry could have been locked and enqueued
+	 * in the ring buffer, so let's flush the ring buffer and try again.
+	 */
+	dequeue_all(head);
+
+	/*
+	 * If the entry was assigned as the LRU entry, let's assign it again,
+	 * because the LRU entry most likely has been changed in dequeue_all().
+	 */
+	if (is_LRU) {
+		entry = TAILQ_FIRST(&head->first);
+		ptr_entry = entry->ptr_entry;
+	} else {
+		entry = *ptr_entry;
+	}
+
+	/* try to lock the entry the second time */
+	if (entry != NULL && util_bool_compare_and_swap64(ptr_entry,
+								entry, NULL))
+		goto evict_found_entry;
+
+	/* the second try failed */
 
+	if (!is_LRU) {
```

Comment:
According to CodeCov all code starting from this line to evict_found_entry label is not covered by tests...

Reply:
Not all code. I do not know what report you are referring to. Please see the latest one:
https://codecov.io/gh/pmem/vmemcache/src/8cf188ff85f6531d22f4ab9a5891f6d889553dc5/src/vmemcache_repl.c
from the latest build:
https://travis-ci.org/pmem/vmemcache/jobs/511398954

Anyway the coverage of this code is random and changes from build to build, because this code covers very rarely happening and very difficult to simulate conditions.

Reply:
This link: https://codecov.io/gh/pmem/vmemcache/pull/165/src/src/vmemcache_repl.c?before=src/vmemcache_repl.c#L387 shows a different picture.

CodeCov is just broken.